message/validation/validation.go

```diff
@@ -118,77 +121,105 @@ func (mv *messageValidator) handlePubsubMessage(pMsg *pubsub.Message, receivedAt
 }
 
 func (mv *messageValidator) handleSignedSSVMessage(signedSSVMessage *spectypes.SignedSSVMessage, topic string, receivedAt time.Time) (*queue.DecodedSSVMessage, error) {
+	decodedMessage := &queue.DecodedSSVMessage{
+		SignedSSVMessage: signedSSVMessage,
+	}
+
 	if err := mv.validateSignedSSVMessage(signedSSVMessage); err != nil {
-		return nil, err
+		return decodedMessage, err
 	}
 
-	if err := mv.validateSSVMessage(signedSSVMessage.SSVMessage, topic); err != nil {
-		return nil, err
+	decodedMessage.SSVMessage = signedSSVMessage.SSVMessage
+
+	if err := mv.validateSSVMessage(signedSSVMessage.SSVMessage); err != nil {
+		return decodedMessage, err
 	}
 
-	committee, validatorIndices, err := mv.getCommitteeAndValidatorIndices(signedSSVMessage.SSVMessage.GetID())
+	committeeData, err := mv.getCommitteeAndValidatorIndices(signedSSVMessage.SSVMessage.GetID())
 	if err != nil {
```

Comment:
this would be a quite heavy operation to perform on every incoming message (imagine constructing slices of 500 indices at a rate of dozens/hundreds of times per second)

i would not rush to optimize this now, but we should definitely do so before finalizing Alan message validation

caching would not be ideal, as the validator indices may change, instead we can leverage the ValidatorStore to keep track of committees' indices and return them in Committee methods (which already return a Committee struct that we should add an Indices filter to)

the reason the ValidatorStore approach is faster (if done correctly) is because it listens to added/removed shares and doesn't have to recompute everything on each method call

Reply:
left a comment